### PR TITLE
Integration test for watching configmap

### PR DESCRIPTION
**Pull request by tomahawk28**
Adding additional integration testing for watching configmap feature. 

## Checklist

- [ ] Update changelog in CHANGELOG.md.

### CODE REVIEW

File: integration/test/watcher/configmap/configmap_test.go [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+		_, err = config.K8sClients.G8sClient().ApplicationV1alpha1().Apps(key.Namespace()).Create(ctx, appCR, metav1.CreateOptions{})
+		if err != nil {
+			t.Fatalf("expected %#v got %#v", nil, err)
+		}
+
+		config.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("creating %#q app cr", key.TestAppReleaseName()))
+	}
+
+	{
+		config.Logger.LogCtx(ctx, "level", "debug", "message", "waiting until user configmap being labelled")
```
Comment (rossf7) [inline integration/test/watcher/configmap/configmap_test.go:155]:
```suggestion
		config.Logger.LogCtx(ctx, "level", "debug", "message", "waiting until user configmap is labelled")
```

Thanks for using 🇬🇧 spelling! ❤️

File: integration/test/watcher/configmap/configmap_test.go [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+			config.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("failed to get a label; retrying in %d", t), "stack", fmt.Sprintf("%v", err))
+		}
+
+		b := backoff.NewMaxRetries(5, backoff.ShortMaxInterval)
+		err := backoff.RetryNotify(o, b, n)
+		if err != nil {
+			t.Fatalf("expected %#v got %#v", nil, err)
+		}
+
+		config.Logger.LogCtx(ctx, "level", "debug", "message", "waited until user configmap being labelled")
```
Comment (rossf7) [inline integration/test/watcher/configmap/configmap_test.go:180]:
```suggestion
		config.Logger.LogCtx(ctx, "level", "debug", "message", "waited until user configmap was labelled")
```

File: integration/test/watcher/configmap/configmap_test.go [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+	}
+
+	{
+		config.Logger.LogCtx(ctx, "level", "debug", "message", "waiting until appcatalog configmap being labelled")
```
Comment (rossf7) [inline integration/test/watcher/configmap/configmap_test.go:184]:
```suggestion
		config.Logger.LogCtx(ctx, "level", "debug", "message", "waiting until appcatalog configmap is labelled")
```

File: integration/test/watcher/configmap/configmap_test.go [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+			config.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("failed to get a label; retrying in %d", t), "stack", fmt.Sprintf("%v", err))
+		}
+
+		b := backoff.NewMaxRetries(5, backoff.ShortMaxInterval)
+		err := backoff.RetryNotify(o, b, n)
+		if err != nil {
+			t.Fatalf("expected %#v got %#v", nil, err)
+		}
+
+		config.Logger.LogCtx(ctx, "level", "debug", "message", "waited until appcatalog configmap being labelled")
```
Comment (rossf7) [inline integration/test/watcher/configmap/configmap_test.go:209]:
```suggestion
		config.Logger.LogCtx(ctx, "level", "debug", "message", "waited until appcatalog configmap was labelled")
```

File: integration/test/watcher/configmap/configmap_test.go [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+	}
+
+	var updatedResourceVersion string
+	{
+		config.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("editing configmap %#q in namespace %#q", key.UserConfigMapName(), key.Namespace()))
```
Comment (rossf7) [inline integration/test/watcher/configmap/configmap_test.go:214]:
```suggestion
		config.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("updating values in configmap %#q in namespace %#q", key.UserConfigMapName(), key.Namespace()))
```

The user will edit the values but it reads a bit off for me. WDYT to this?

File: integration/test/watcher/configmap/configmap_test.go [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+
+		cm.Data["values"] = "test: userconfigmap"
+		updatedCM, err := config.K8sClients.K8sClient().CoreV1().ConfigMaps(key.Namespace()).Update(ctx, cm, metav1.UpdateOptions{})
+		if err != nil {
+			t.Fatalf("expected %#v got %#v", nil, err)
+		}
+
+		updatedResourceVersion = updatedCM.GetResourceVersion()
+
+		config.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("edited configmap %#q in namespace %#q", key.UserConfigMapName(), key.Namespace()))
```
Comment (rossf7) [inline integration/test/watcher/configmap/configmap_test.go:229]:
```suggestion
		config.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("updated values in configmap %#q in namespace %#q", key.UserConfigMapName(), key.Namespace()))
```

File: integration/test/watcher/configmap/configmap_test.go [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+	}
+
+	versionAnnotation := fmt.Sprintf("%s/%s", annotation.AppOperatorPrefix, annotation.LatestConfigMapVersion)
+
+	{
+		config.Logger.LogCtx(ctx, "level", "debug", "message", "waiting until app CR annotate by user configmap's resourceVersion")
```
Comment (rossf7) [inline integration/test/watcher/configmap/configmap_test.go:235]:
```suggestion
		config.Logger.LogCtx(ctx, "level", "debug", "message", "waiting until app CR is annotated with user configmap's resourceVersion")
```

File: integration/test/watcher/configmap/configmap_test.go [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+			config.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("failed to get an annotation; retrying in %d", t), "stack", fmt.Sprintf("%v", err))
+		}
+
+		b := backoff.NewMaxRetries(5, backoff.ShortMaxInterval)
+		err := backoff.RetryNotify(o, b, n)
+		if err != nil {
+			t.Fatalf("expected %#v got %#v", nil, err)
+		}
+
+		config.Logger.LogCtx(ctx, "level", "debug", "message", "waited until app CR annotate by user configmap's resourceVersion")
```
Comment (rossf7) [inline integration/test/watcher/configmap/configmap_test.go:262]:
```suggestion
		config.Logger.LogCtx(ctx, "level", "debug", "message", "waited until app CR is annotated with user configmap's resourceVersion")
```

File: integration/test/watcher/configmap/configmap_test.go [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+		err = config.K8sClients.G8sClient().ApplicationV1alpha1().Apps(key.Namespace()).Delete(ctx, key.TestAppReleaseName(), metav1.DeleteOptions{})
+		if err != nil {
+			t.Fatalf("expected %#v got %#v", nil, err)
+		}
+
+		config.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleted app CR %#q", key.TestAppReleaseName()))
+	}
+
+	{
+		config.Logger.LogCtx(ctx, "level", "debug", "message", "waiting until watching label get deleted")
```
Comment (rossf7) [inline integration/test/watcher/configmap/configmap_test.go:327]:
```suggestion
		config.Logger.LogCtx(ctx, "level", "debug", "message", "waiting until watching label is deleted")
```

File: integration/test/watcher/configmap/configmap_test.go [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+			config.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("still get label; retrying in %d", t), "stack", fmt.Sprintf("%v", err))
+		}
+
+		b := backoff.NewMaxRetries(5, backoff.ShortMaxInterval)
+		err := backoff.RetryNotify(o, b, n)
+		if err != nil {
+			t.Fatalf("expected %#v got %#v", nil, err)
+		}
+
+		config.Logger.LogCtx(ctx, "level", "debug", "message", "waited until watching label get deleted")
```
Comment (rossf7) [inline integration/test/watcher/configmap/configmap_test.go:352]:
```suggestion
		config.Logger.LogCtx(ctx, "level", "debug", "message", "waited until watching label was deleted")
```

File: integration/test/watcher/configmap/configmap_test.go [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+
+			if _, ok := cm.GetLabels()[pkglabel.Watching]; ok {
+				return microerror.Maskf(testError, fmt.Sprintf("%#q label still found", pkglabel.Watching))
+			}
+
+			return nil
+		}
+
+		n := func(err error, t time.Duration) {
+			config.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("still get label; retrying in %d", t), "stack", fmt.Sprintf("%v", err))
```
Comment (rossf7) [inline integration/test/watcher/configmap/configmap_test.go:343]:
```suggestion
			config.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("still getting label; retrying in %d", t), "stack", fmt.Sprintf("%v", err))
```

File: integration/test/watcher/configmap/configmap_test.go [hunk 12]
```diff
@@ -0,0 +1,355 @@
...
+// - Create app CR and wiring user configmap and appcatalog
+//
+// - Update user configmap and check the latest resource version is set on the annotation
+//   of app CR.
+//
+// - Update appcatalog onfigmap and check the latest resource version is set on the annotation
+//   of app CR.
+//
+//
+// - Delete app CR and check the watching label is deleted.
```
Comment (rossf7) [inline integration/test/watcher/configmap/configmap_test.go:36]: Super nice and thanks for writing up the steps. Makes it easier to figure out what the test is doing!

**Review by rossf7 — COMMENTED** [review]
Very nice and thanks for extending the coverage to include the watcher.

Test steps look good just some suggestions on the logging.

File: integration/test/watcher/configmap/configmap_test.go (hunk 5 above)
Comment (tomahawk28) [reply integration/test/watcher/configmap/configmap_test.go:214]: Nicer than editing I think, thanks for a suggestion!

**Comment by tomahawk28** [comment]
No need to update changelog since it's integration test, @rossf7 PTAL :)

**Comment by tomahawk28** [comment]
going with a ping

**Comment by rossf7** [comment]
Late to the party but LGTM. Thx :)

**Pull request merged**